### PR TITLE
md5: add assembly implementation for loongarch64

### PR DESCRIPTION
**Pull request by zhoumin2**
This change can improve md5 performance by using a hand-optimized assembly implementation of the inner loop of md5 calculation. This implementation refered to md5-x86_64.pl and made more effort to reorder instructions for separating data dependencies as much as possible.
```bash
Test with:
$ openssl speed md5

3A5000
type             16 bytes    64 bytes     256 bytes    1024 bytes   8192 bytes   16384 bytes
md5              45061.04k   130440.75k   291105.28k   421101.23k   484639.27k   488320.43k
md5-modified     47179.95k   139015.57k   308836.69k   445963.26k   512540.67k   518215.00k
                   +5%         +7%          +6%          +6%          +6%          +6%

3A6000
type             16 bytes    64 bytes     256 bytes    1024 bytes   8192 bytes   16384 bytes
md5              60070.06k   161822.76k   325817.60k   438017.02k   486864.21k   492243.31k
md5-modified     62827.74k   170294.04k   343795.03k   463324.50k   515831.13k   520060.93k
                   +5%         +5%          +6%          +6%          +6%          +6%
```

### CODE REVIEW

**Comment by zhoumin2** [comment]
@shipujin , please help me review this patch, thanks.

**Comment by shipujin** [comment]
> @shipujin , please help me review this patch, thanks.

I'm not too familiar with the underlying assembly, waiting for others in the community to review

**Review by paulidale — APPROVED** [review]

**Comment by zhoumin2** [comment]
Kindly ping.
Hope for a second review for this change, thanks.

**Comment by zhoumin2** [comment]
Kindly ping @openssl/committers for the second review, thanks.

**Comment by zhoumin2** [comment]
Hi, @t8m , could you give me a review for this change? Thanks very much.

**Comment by t8m** [comment]
I am sorry but my personal opinion (not the opinion of the project) always was that we should not add new assembler implementations for algorithms in the legacy provider. So that is why I will not approve this PR. That of course does not prevent anybody else from OpenSSL committers approving it.

**Comment by t8m** [comment]
> I'd also would raise an OTC hold to discuss whether we are willing to add assembler implementations of legacy algs.

We have discussed this and there were, as far as I remember, mixed opinions but the majority of OTC wasn't against adding new assembler for legacy algorithms.

**Comment by paulidale** [comment]
> …. we should not add new assembler implementations for algorithms in the legacy provider.

Noting that MD5 is **not** in the legacy provider.  It's in the default provider.

**Review by paulidale — APPROVED** [review]

**Comment by zhoumin2** [comment]
> I can formally approve it (I don't share @t8m opinion) but I don't understand the code. 

There exists heavy data dependencies in the inner loop of md5 calculation. The assembly implementation has a chance to perform hand optimization. This implementation is highly similar to md5-x86_64.pl which also achieved a 6% improvement in performance.

The actual assembler instructions indeed have differences. The detailed information about LoongArch ISA can be found on the following page: https://loongson.github.io/LoongArch-Documentation/LoongArch-Vol1-EN.html. 

I'm also pleased to answer the questions about the implementation details.

**Comment by paulidale** [comment]
I just checked on qemu on our CI machines and it is too old to support Loongarch.

**Comment by paulidale** [comment]
I didn't doubt that the code worked.  Crypto code like this is pretty easy to show general correctness -- it either works or it produces garbage.

I do think it would be good, nonetheless, to have a Loongarch build or two in our CI loops.  If the qemu the runners used supported it, I'd have added it.

**Comment by t8m** [comment]
**OTC: As this is not an algorithm in the legacy provider, there is no need to block this PR.**

**Comment by zhoumin2** [comment]
Can I do anything to let the review continue?

**Comment by paulidale** [comment]
This needs a second approval from @openssl/committers

**Comment by zhoumin2** [comment]
> This needs a second approval from @openssl/committers

Thanks, Pauli.

Kindly ping @openssl/committers for a second approval for this change.

**Comment by zhoumin2** [comment]
Kindly ping @openssl/committers for a second approval for this change.

**Comment by zhoumin2** [comment]
Kindly ping @openssl/committers for a second review.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:20]:
```suggestion
```

Let's remove this line because no saved registers are used.  This line made me doubt "hey, don't tell me a saved register is used w/o saving and restoring [again](https://github.com/openssl/openssl/pull/22817)"!

File: crypto/md5/md5_local.h
Comment (xry111) [inline crypto/md5/md5_local.h:19]:
```suggestion
     defined(_M_X64) || defined(__aarch64__) || (defined(__loongarch__) && __loongarch_grlen == 64)
```

If I read the code correctly, it requires 64-bit GPR but does not require LP64 ABI (i.e. it should works with ILP32 ABI too).  If it requires LP64 (LP64S/LP64F/LP64D) ABI, just test `defined(__loongarch_lp64)`.

`__loongarch64` is obsolete because it actively confuses people to mix up "LP64 ABI" and "64-bit GPR".  These are two different concepts.  64-bit GPR is a hardware feature, but LP64 ABI is a software interface.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:20]: Thanks for the reminder. This line can be removed.

File: crypto/md5/md5_local.h
Comment (zhoumin2) [reply crypto/md5/md5_local.h:19]: Yes. Thanks for your review.  Using ```__loongarch_grlen``` is more appropriate.

**Comment by zhoumin2** [comment]
Kindly ping @openssl/committers for a second approval for this change.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:18]: one of the longarch documents references a0-a7 as parameter passing registers.. Is r4-r11 actually defined as param passing somewhere in the spec?

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:34]: Maybe comment what this is.. It looks like a sign extended version of T_i_h

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:36]: This is 64bit? Cant see how all 64 bits of this get used further down (they seem to operate on words)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:40]: These immediate values of 12 and 20 bits are very strange instructions :)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:50]: cyclic left shift by cycling right :)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:169]: any reason not to do beqz here? (it is a size_t)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [inline crypto/md5/asm/md5-loongarch64.pl:189]: Are these aliases.. Couldnt see them in the spec for some reason

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [reply crypto/md5/asm/md5-loongarch64.pl:18]: https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#general-purpose-registers

a0-a7 are aliases of r4-r11.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [reply crypto/md5/asm/md5-loongarch64.pl:17]: https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#general-purpose-registers

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [reply crypto/md5/asm/md5-loongarch64.pl:189]: `move a, b` is an alias of `or a, b, $r0`.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [reply crypto/md5/asm/md5-loongarch64.pl:40]: Basically to load a 32-bit immediate imm we use `lu12i.w reg, imm >> 12` and `ori reg, reg, imm & 0xfff`.

For readability we can use an assembler macro `li.w reg, imm` which will be expanded to lu12i.w + ori.  Here the author is doing some instruction scheduling as a micro-optimization, thus the lu12i.w and ori instructions are explicitly written and an unrelated instruction is "scheduled" in between them.

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
This still needs a second approval.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [reply crypto/md5/asm/md5-loongarch64.pl:36]: As we are still waiting for another approval, I'd like to fix this because using ld.d on an address not aligned to 8-byte boundary may cause a SIGBUS on low-end hardware lacking unaligned access support.

Per a private communication with the author, he may be slow to respond as at now for some reason.

Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:36]:
```suggestion
    $code .= "	ld.w	$t0,$a1,0		/* (NEXT STEP) X[0] */\n" if ($pos == -1);
```

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:45]:
```suggestion
	ld.w	$t0,$a1,$k_next*4		/* (NEXT STEP) X[$k_next] */
```

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:79]:
```suggestion
	ld.w	$t0,$a1,$k_next*4		/* (NEXT STEP) X[$k_next] */
```

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:110]:
```suggestion
	ld.w	$t0,$a1,$k_next*4		/* (NEXT STEP) X[$k_next] */
```

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:144]:
```suggestion
        $code .= "	ld.w	$t0,$a1,$k_next*4		/* (NEXT STEP) X[$k_next] */\n";
```

File: crypto/md5/asm/md5-loongarch64.pl
Comment (xry111) [inline crypto/md5/asm/md5-loongarch64.pl:169]:
```suggestion
	beqz	$a2,.Lend		# cmp nbr with 0, jmp if nbr == 0
```

**Review by xry111 — CHANGES_REQUESTED** [review]
I've tested these changes locally.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:18]: Yes, thanks!

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:17]: As @xry111 pointed out, we can find out the meanings of all the 32 registers from the above link.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (slontis) [reply crypto/md5/asm/md5-loongarch64.pl:40]: That is the kind of thing that would be good as a comment in the code :)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:34]: Yes. Here in md5, we need write a 32-bit constant into a general register, such as the constant of ```0xd76aa478``` in the round 1 and step 1. In LoongArch we have to use two instructions of ```lu12i.w``` and ```ori``` to finish this work. Meanwhile, the instruction ```lu12i.w``` treats the 20-bit immediate as a signed number. So if the ```T_i_h``` is greater than ```1^19```, we need provide ```lu12i.w``` a corresponding negative number whose complement equals to the sign extension of ```T_i_h```.

The details of the instruction ```lu12i.w``` can be found as following:
[https://loongson.github.io/LoongArch-Documentation/LoongArch-Vol1-EN.html#_lu12i_w_lu32i_d_lu52i_d](https://loongson.github.io/LoongArch-Documentation/LoongArch-Vol1-EN.html#_lu12i_w_lu32i_d_lu52i_d)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:40]: I will add some comments in the code.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:36]: Thanks for the reminder. Loading as 32-bit is enough.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:50]: Yes, LoongArch has not the instruction for directly cyclic left shifting.

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl:189]: I'm sorry, I didn't find the public spec for the `move` instruction.

The ```move``` instruction is probably handled by the ```binutils gas```. There are some useful references in the project ```binutils-gdb``` for this question as following:
[https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=opcodes/loongarch-opc.c#l349](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=opcodes/loongarch-opc.c#l349)
[https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=opcodes/loongarch-opc.c#l368](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=opcodes/loongarch-opc.c#l368)

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl]: Thanks!

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl]: Thanks!

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl]: Thanks!

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl]: Thanks!

File: crypto/md5/asm/md5-loongarch64.pl
Comment (zhoumin2) [reply crypto/md5/asm/md5-loongarch64.pl]: Thanks.

**Comment by slontis** [comment]
Looks good, The only other question I have is, do any registers need to be saved/restored on enter/exit.. (Including the top 32 bits)

**Comment by zhoumin2** [comment]
> Looks good, The only other question I have is, do any registers need to be saved/restored on enter/exit.. (Including the top 32 bits)

There are eleven general purpose registers needed to be saved/restored on enter/exit, including: ```$r3``` (alias ```$sp```), ```$r22``` (alias ```$fp / $s9```) and ```$r23 - $r31``` (alias ```$s0 - $s8```). In this assembly implementation we doesn't use these registers, so it's no need to save/restore other registers.

The general purpose register convention can be found as following:
[https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#general-purpose-registers](https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#general-purpose-registers)

**Comment by slontis** [comment]
reapproved.

**Comment by slontis** [comment]
ping For second review.

**Comment by zhoumin2** [comment]
Kindly ping @openssl/committers for a second approval.

**Comment by zhoumin2** [comment]
Kindly ping @openssl/committers for a second review.

**Comment by slontis** [comment]
@t8m Would you be able to quick review. I did a fairly thorough review (looked at both the RFC and learned the assembler instructions).

**Review by t8m — APPROVED** [review]
It is a little bit against my mind to approve new assembler implementations for these legacy insecure algorithms but anyway.

**Comment by xry111** [comment]
@zhoumin2 I guess you may need to remove the checklist in the desc if we don't plan to complete it.

**Comment by zhoumin2** [comment]
> @zhoumin2 I guess you may need to remove the checklist in the desc if we don't plan to complete it.

Yes, thank you for your suggestion. 
There is no extra documentation to add or update, and the added perl script has some useful comments. The tests for MD5 are ready-made and no need to add or update.

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**
